### ISSUE TITLE
Error on the code (?)

### ISSUE THREAD

**Issue by Sulwon-0516** [shihui2010/symbolic_simplifier#3]
I'm trying to run your code, but I had error so I write this issues.

At EncoderDecoder.py

~~~python
for ne in unequals:
    ne = ne.clone()
    token_tree_to_id_tree(ne, vocab)
    self.enforce_encoding(s2=[ne], encoded1=hidden_states, eos=eos, positive=False)
~~~

Look at the **equivalences** and **unequals**, which are from the result of beam-search.
We want to optimize the model to embed closer when input is from **equivalences** (or more far away from **unequals**) with **self.enforce_encoding()**

In the **self.enforce_encoding()**, it back-propagate through encoder and modify their weights and values.

As you can see above, this code reuse **hidden_states** for all **unequals** and **equivalences**.
It leads to the pytorch raise Error when self.enforce_encoding is called more than one time. 
-> (it can't back-prop the **hidden_states** because the weights are modified.)

I think, if my understanding is right,  changing the code like this would be optimal.

~~~python
for ne in unequals:
    ne = ne.clone()
    token_tree_to_id_tree(ne, vocab)
    self.enforce_encoding(s1=sample_input, s2=[ne], eos=eos, positive=False)
~~~

It will not make issues, because it calculate the original embedding every time, with modified encoder.
Or, change the code to accumulate the backward gradients and calling the optimizer.step() only one time per each step.

Also, I have another question.

Inside the enforce_encoding....

~~~python
encoder_outputs1 = self.encode(s1, eos)[2] if encoded1 is None else encoded1
encoder_outputs2 = self.encode(s2, eos)[2] if encoded2 is None else encoded2

loss = torch.nn.MSELoss()(encoder_outputs1[0][:self._hps["hidden_unit_split"]],
                            encoder_outputs2[0][:self._hps["hidden_unit_split"]]) * self._hps["l2_weight"]
~~~

we use the first return value of **self.encode()**.
It's **encoder_outputs**, and it's different from **hidden_states** isn't it?
Are they are same? or

Am I right? 
Yet, I didn't read the full-code perfectly, so maybe there would some my mis-understanding. 

Thank you. 
Stay safe from Covid and have nice day. 

P.S it seems this code would work well before torch 1.5. (not tested)
